PIPS/pip-1.md

```diff
@@ -0,0 +1,146 @@
+    address assetHash = _msgSender();
+
+    // `hash` is any supported hashing function of the respective blockchain
+    key = hash(assetHash, nativeChainId, nativeLockProxy, nativeAssetHash);
+
+    require(registry[key] != true);
+    require(balances[key] == 0);
+    require(getBalanceFor(assetHash) == delegatedSupply);
+
+    registry[key] = true;
+    balances[key] = delegatedSupply;
+
+    RegisterAssetTxArgs memory txArgs = RegisterAssetTxArgs({
+        assetHash: assetHash,
+        nativeAssetHash: nativeAssetHash
+    });
+
+    bytes memory txData = _serializeRegisterAssetTxArgs(txArgs);
+    require(eccm.crossChain(nativeChainId, nativeLockProxy, "registerAsset", txData), "EthCrossChainManager crossChain executed error!");
+
+    emit DelegateAsset(assetHash, nativeChainId, nativeLockProxy, nativeAssetHash);
+}
+```
+
+The `registerAsset` is called by cross-chain transactions from other LockProxies.
+The purpose of this function is to update the `registry` mapping which will be checked when a user calls the `lock` function.
+This helps to ensure that the user does not call `lock` with incorrect parameters.
+
+In the LockProxy, `registerAsset` can be implemented as:
+```
+function registerAsset(bytes memory argsBs, bytes memory fromContractAddr, uint64 fromChainId) onlyManagerContract public {
+  TxRegisterAssetArgs memory args = _deserializTxRegisterAssetArgs(argsBs);
+
+  // `hash` is any supported hashing function
+  key = hash(args.nativeAssetHash, fromChainId, fromContractAddr, args.assetHash);
+
+  require(registry[key] != true);
+  registry[key] = true;
+}
+```
+
+`setManagerProxy` can be restricted to be called once. After it is called the first time, the `managerProxyContract` address cannot be changed.
+
+`bindProxyHash` and `bindAssetHash` functions can be removed.
+
+The `lock` function can be modified to require the following parameters:
+- address fromAssetHash
+- uint64 toChainId
+- bytes memory targetProxyHash
+- bytes memory toAssetHash
+- bytes memory toAddress
+- uint256 amount
+
+The `balances` mapping should be updated within the `lock` function:
+```
+bytes32 key = hash(fromAssetHash, toChainId, targetProxyHash, toAssetHash);
+require(registry[key] == true);
+
+balances[key] += amount;
+
+// transfer tokens from user to LockProxy
+
+address fromContractAddr = address(this);
+TxArgs memory txArgs = TxArgs({
+    fromContractAddr: fromContractAddr,
+    fromAssetHash: fromAssetHash,
+    toAssetHash: toAssetHash,
+    toAddress: toAddress,
+    amount: amount
+});
+bytes memory txData = _serializeTxArgs(txArgs);
+
+require(eccm.crossChain(toChainId, targetProxyHash, "unlock", txData), "EthCrossChainManager crossChain executed error!");
+
+emit LockEvent(address fromAssetHash, address fromAddress, uint64 toChainId, bytes toAssetHash, bytes toAddress, uint256 amount, ...optional);
+```
+
+The `unlock` function should be modified to require the following parameters:
```

Comment:
whether add a explain this method is only can be called by manager contract?

Reply:
ok will add it 👍